File: C++/DesignPattern/CreationalDesignPattern/ObserverDesign_pattern.cpp
#include <iostream>
#include <memory>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

struct Subject {
    virtual void addObserver(struct Observer* observer) = 0;
    virtual void removeObserver(struct Observer* observer) = 0;
    virtual ~Subject() = default;
};

struct Follower {
    virtual void notify(const string& message) = 0;
    virtual ~Follower() = default;
};

class WhatsAppBroadCast {
    vector<shared_ptr<Follower>> followers;

public:
    void addFollower(const shared_ptr<Follower>& follower) {
        followers.push_back(follower);
    }

    void removeFollower(const shared_ptr<Follower>& follower) {
        followers.erase(remove(followers.begin(), followers.end(), follower), followers.end());
    }

    void sendMessage(const string& message) {
        for (auto& follower : followers) {
            follower->notify(message);
        }
    }
};

struct Prateek : Follower {
    void notify(const string& message) override {
        cout << " Prateek Received Message: " << message << "\n";
    }
};

struct Abhinav : Follower {
    void notify(const string& message) override {
        cout << "Abhinav Received Message: " << message << "\n";
    }
};

int main() {
    WhatsAppBroadCast whatsAppBroadCast;
    whatsAppBroadCast.addFollower(make_shared<Prateek>());
    whatsAppBroadCast.addFollower(make_shared<Abhinav>());

    whatsAppBroadCast.sendMessage("pradeep is teaching Observer pattern");
    return 0;
}

File: Python/DesignPattern/CreationalDesignPattern/ObserverDesign_pattern.py
from abc import ABC, abstractmethod


class Subject(ABC):
    @abstractmethod
    def add_observer(self, observer): pass

    @abstractmethod
    def remove_observer(self, observer): pass


class Follower(ABC):
    @abstractmethod
    def notify(self, message: str): pass


class WhatsAppBroadCast:
    def __init__(self):
        self.followers = []

    def add_follower(self, follower: Follower):
        self.followers.append(follower)

    def remove_follower(self, follower: Follower):
        self.followers.remove(follower)

    def send_message(self, message: str):
        for follower in self.followers:
            follower.notify(message)


class Prateek(Follower):
    def notify(self, message: str):
        print(f" Prateek Received Message: {message}")


class Abhinav(Follower):
    def notify(self, message: str):
        print(f"Abhinav Received Message: {message}")


if __name__ == "__main__":
    whatsapp_broadcast = WhatsAppBroadCast()
    whatsapp_broadcast.add_follower(Prateek())
    whatsapp_broadcast.add_follower(Abhinav())

    whatsapp_broadcast.send_message("pradeep is teaching Observer pattern")
